MapboxNavigation/SimulatedRoute.swift

```diff
@@ -5,85 +5,198 @@ protocol SimulatedRouteDelegate : class {
+        
+        for (i, location) in locations.enumerated() {
+            if i > 0 {
+                let distance = location.distance(from: processedLocations.last!)
+                let time = distance / location.speed
+                let newLocation = location.shifted(to: processedLocations.last!.timestamp.addingTimeInterval(time))
+                processedLocations.append(newLocation)
+            } else {
+                processedLocations.append(location.shifted(to: Date()))
+            }
+        }
+        
+        return processedLocations
+    }
+    
+    // Calculate speed based on turn penalty.
+    fileprivate func calculateSpeed(for locations: [SimulatedLocation]) -> [SimulatedLocation] {
+        var processedLocations = [SimulatedLocation]()
         
-        let totalDistance = distance(along: polyline)
-        let stops = UInt(totalDistance / distanceFilter)
-        let distancePerStop = totalDistance / Double(stops)
+        for (i, location) in locations.enumerated() {
+            guard i < locations.count else { break }
+            let reversedTurnPenalty = maximumTurnPenalty - locations[i].turnPenalty
+            let speed = reversedTurnPenalty.scale(minimumIn: minimumTurnPenalty, maximumIn: maximumTurnPenalty, minimumOut: minimumSpeed, maximumOut: maximumSpeed)
+            let processedLocation = location.withSpeed(speed)
+            processedLocations.append(processedLocation)
+        }
         
-        for index in 0...stops {
-            let distance = CLLocationDistance(UInt(distanceFilter) * (index + 1))
-            var course: CLLocationDegrees = 0
-            let newCoordinate = coordinate(at: distance, fromStartOf: polyline)
-            guard let currentCoordinate = newCoordinate else { continue }
+        return processedLocations
+    }
+    
+    // Insert new locations to smooth out gaps longer than 1 second.
+    fileprivate func interpolate(locations: [SimulatedLocation]) -> [SimulatedLocation] {
+        var processedLocations = [SimulatedLocation]()
+        
+        for (i, location) in locations.enumerated() {
+            processedLocations.append(location)
             
-            if let nextCoordinate = coordinate(at: distance + distanceFilter, fromStartOf: polyline) {
-                course = currentCoordinate.direction(to: nextCoordinate)
+            if i < locations.count-1 {
+                let speed = (maximumSpeed - minimumSpeed) / 2
+                let nextLocation = locations[i+1]
+                let distance = floor(location.distance(from: nextLocation))
+                let timeToTravel = distance / speed
+                
+                if timeToTravel > 1 {
+                    let newCoordinate = coordinate(at: speed, fromStartOf: [location.coordinate, nextLocation.coordinate])
+                    let newLocation = SimulatedLocation(coordinate: newCoordinate!,
+                                                        altitude: 0,
```

Comment:
Just in case we decided to do something with altitude, say for overlapping over surface streets and highways, maybe this should be changed to:

```
altitude: location.altitude,
```